### PR TITLE
Feature groups ranking

### PR DESCRIPTION
**Pull request by vpontis**

### CODE REVIEW

**Comment by vpontis** [comment]
dont merge yet. i'm just doing this to keep track of progress

**Comment by vpontis** [comment]
check this out now @stubailo

**Comment by stubailo** [comment]
Looking at it

File: app/models/game.rb [hunk 1]
```diff
@@ -74,25 +74,32 @@ def users_are_unique
...
+    loser_ratings = self.losers.map { |loser| loser.points }
 
-    q_winners = 10**(winners_rating/400)
-    q_losers = 10**(losers_rating/400)
-
-    expected_winners = q_winners/(q_winners + q_losers)
-    expected_losers = q_losers/(q_winners + q_losers)
-
-    point_change = 32*(1 - expected_winners)
+    point_change = Game.point_change(winner_ratings, loser_ratings)
```
Comment (stubailo) [inline app/models/game.rb:80]: I really want to have different point gains for the two players, but maybe we can do that later

File: app/models/game.rb [hunk 2]
```diff
@@ -74,25 +74,32 @@ def users_are_unique
...
 
     self.losers.each do |loser|
-      loser.points -= (point_change * (loser.points)/losers_rating)
+      loser.points -= point_change 
       loser.save
     end
   end
+
+  def self.point_change(winner_ratings, loser_ratings)
+    winners_rating = (winner_ratings[0] + winner_ratings[1])/2
```
Comment (stubailo) [inline app/models/game.rb]: winner_ratings.sum is better here

File: app/models/game.rb (hunk 1 above)
Comment (vpontis) [inline app/models/game.rb:80]: Yea, we need to talk about how to do this.

File: app/models/group.rb [hunk 3]
```diff
@@ -6,10 +6,42 @@ class Group < ActiveRecord::Base
...
 
+  after_save :calculate_player_rankings
+  after_update :calculate_player_rankings
+
   def player_usernames
     self.players.map {|p| p.username}
   end
 
+  def games
+    group_games = Game.all.select {|game| self.game_in_group?(game)}
```
Comment (stubailo) [inline app/models/group.rb:17]: This is pretty shitty, we should at least make an issue to make it more efficient

File: app/models/group.rb [hunk 4]
```diff
@@ -6,10 +6,42 @@ class Group < ActiveRecord::Base
...
+    self.group_players.each { |p| p.points = 1000; p.save; }
+
+    self.games.each do |game|
+      group_game_winners = game.winners.map {|winner| self.group_players.find_by(player_id: winner.id)}
+      group_game_losers = game.losers.map {|loser| self.group_players.find_by(player_id: loser.id)}
+
+      winner_ratings = group_game_winners.map {|g| g.points }
+      loser_ratings = group_game_losers.map {|g| g.points }
+
+      point_change = Game.point_change(winner_ratings, loser_ratings)
```
Comment (stubailo) [inline app/models/group.rb:32]: why isn't point_change just a method on a game instance?

File: app/models/group.rb [hunk 5]
```diff
@@ -6,10 +6,42 @@ class Group < ActiveRecord::Base
...
+      
+      group_game_winners.each do |winner|
```
Comment (stubailo) [inline app/models/group.rb:34]: This is definitely duplicated code from above, DRY?

File: app/models/player.rb [hunk 6]
```diff
@@ -48,8 +48,8 @@ def ranking
   end
 
   def games_in_group(group)
-    group_wins = wins.filter {|win| all_players_in_group(win, group)}
-    group_losses = losses.filter {|loss| all_players_in_group(loss, group)}
+    group_wins = wins.filter {|win| group.game_in_group?(win)}
+    group_losses = losses.filter {|loss| group.game_in_group?(loss)}
     return group_wins + group_losses
```
Comment (stubailo) [inline app/models/player.rb]: These won't be ordered properly

File: app/models/player.rb [hunk 7]
```diff
@@ -65,6 +65,11 @@ def unconfirmed_games
     return unconfirmed.sort {|a,b| a.created_at <=> b.created_at}
   end
 
+  def group_points(group)
+    group.save
+    group_player = GroupPlayer.find_by(player_id: self.id)
```
Comment (stubailo) [inline app/models/player.rb]: Uh don't you want to find by player_id and group_id?

File: app/models/group.rb (hunk 4 above)
Comment (vpontis) [inline app/models/group.rb:32]: The point change is different depending on what context you are considering the game.

File: app/models/player.rb (hunk 7 above)
Comment (vpontis) [inline app/models/player.rb]: self.id is player_id

File: test/models/group_test.rb [hunk 8]
```diff
@@ -1,7 +1,48 @@
 require 'test_helper'
 
 class GroupTest < ActiveSupport::TestCase
-  # test "the truth" do
-  #   assert true
-  # end
+  test "rankings not updated if players not in group" do
```
Comment (stubailo) [inline test/models/group_test.rb:4]: sick tests bro

File: app/models/game.rb (hunk 1 above)
Comment (stubailo) [inline app/models/game.rb:80]: Yeah don't worry about it for now

File: app/models/group.rb (hunk 4 above)
Comment (stubailo) [inline app/models/group.rb:32]: Makes sense

File: app/models/group.rb (hunk 5 above)
Comment (stubailo) [inline app/models/group.rb:34]: Wait, wtf? I feel like the pull request UI just fucked up

File: app/models/player.rb (hunk 7 above)
Comment (stubailo) [inline app/models/player.rb]: Where is the group_id included

Comment (vpontis) [inline app/models/player.rb]: good point

File: app/models/group.rb (hunk 3 above)
Comment (vpontis) [inline app/models/group.rb:17]: done

**Comment by vpontis** [comment]
@stubailo @dsessoms this should be done. let me know if you want any other changes.

**Pull request merged**